Tasks/AzureRmWebAppDeployment/AzureRmWebAppDeployment.ts

```diff
@@ -0,0 +1,234 @@
+
+        if(virtualApplication) {
+            publishingProfile.destinationAppUrl += "/" + virtualApplication;
+        }
+
+        if(webAppUri) {
+            tl.setVariable(webAppUri, publishingProfile.destinationAppUrl);
+        }
+
+        if(canUseWebDeploy(useWebDeploy)) {
+            await DeployUsingMSDeploy(webDeployPkg, webAppName, publishingProfile, removeAdditionalFilesFlag,
+                            excludeFilesFromAppDataFlag, takeAppOfflineFlag, virtualApplication, setParametersFile,
+                            additionalArguments, isFolderBasedDeployment);
+        } else {
+            tl.debug(tl.loc("Initiateddeploymentviakuduserviceforwebapppackage", webDeployPkg));
+            var azureWebAppDetails = await azureRmUtil.getAzureRMWebAppConfigDetails(SPN, webAppName, resourceGroupName, deployToSlotFlag, slotName);
+            await DeployUsingKuduDeploy(webDeployPkg, azureWebAppDetails, publishingProfile, virtualApplication, isFolderBasedDeployment);
+        }
+    } catch (error) {
+        tl.setResult(tl.TaskResult.Failed, error);
+    }
+}
+
+/**
+ * Executes Web Deploy command
+ * 
+ * @param   webDeployPkg                   Web deploy package
+ * @param   webAppName                      web App Name
+ * @param   publishingProfile               Azure RM Connection Details
+ * @param   removeAdditionalFilesFlag       Flag to set DoNotDeleteRule rule
+ * @param   excludeFilesFromAppDataFlag     Flag to prevent App Data from publishing
+ * @param   takeAppOfflineFlag              Flag to enable AppOffline rule
+ * @param   virtualApplication              Virtual Application Name
+ * @param   setParametersFile               Set Parameter File path
+ * @param   additionalArguments             Arguments provided by user
+ * 
+ */
+async function DeployUsingMSDeploy(webDeployPkg, webAppName, publishingProfile, removeAdditionalFilesFlag, 
+        excludeFilesFromAppDataFlag, takeAppOfflineFlag, virtualApplication, setParametersFile, additionalArguments, isFolderBasedDeployment) {
+
+    var isParamFilePresentInPackage = isFolderBasedDeployment ? false : await msDeployUtility.containsParamFile(webDeployPkg);
+    setParametersFile = getSetParamFilePath(setParametersFile);
+    var msDeployPath = await msDeployUtility.getMSDeployFullPath();
+    var msDeployCmdArgs = msDeployUtility.getMSDeployCmdArgs(webDeployPkg, webAppName, publishingProfile, removeAdditionalFilesFlag,
+        excludeFilesFromAppDataFlag, takeAppOfflineFlag, virtualApplication, setParametersFile, additionalArguments, isParamFilePresentInPackage, isFolderBasedDeployment);
+
+    var isDeploymentSuccess = true;
+    var deploymentError = null;
+    try {
+
+        var msDeployBatchFile = tl.getVariable('System.DefaultWorkingDirectory') + '\\' + 'msDeployCommand.bat';
+        var msDeployCommand = '"' + msDeployPath + '" ' + msDeployCmdArgs;
+        fs.writeFileSync(msDeployBatchFile, msDeployCommand);
```

Comment:
Can you use tl.writeFile() instead of fs.writeFileSync() as writeFile() internally calls fs.wrtieFileSync(). Makes it easier to write tests.